README.md
# cmd2-submenu
This project provides a submenu system for cmd2

## Installing
To install the plugin, do:
```
$ pip install cmd2-submenu
```

## How to use

## Running tests

`cmd2` uses a three tiered testing strategy to test on multiple versions
of python on various platforms. This plugin uses the same strategy:

- [pytest](https://pytest.org) runs the unit tests
- [tox](https://tox.readthedocs.io/) runs the unit tests on multiple versions
  of python
- [AppVeyor](https://www.appveyor.com/) and [TravisCI](https://travis-ci.com)
  run the tests on the various supported platforms

### Running unit tests

Run `pytest` from the top level directory of this plugin to run all the
unit tests.

### Use tox to run unit tests in multiple versions of python

The included `tox.ini` is setup to run the unit tests in python 3.4, 3.5,
and 3.6. In order for `tox` to work, you need to have different versions of
python executables available in your path.
[pyenv](https://github.com/pyenv/pyenv) is one method of doing this easily.
Once `pyenv` is installed, use it to install multiple versions of python:

```
$ pyenv install 3.4.8
$ pyenv install 3.5.5
$ pyenv install 3.6.5
$ pyenv local 3.6.5 3.5.5 3.4.8
```

This will create a `.python-version` file and instruct the `pyenv` shims
to make `python3.6`, `python3.5`, and `python3.4` launch the appropriate
versions of python.

Once these executables are configured, invoking `tox` will create a
virtual environment for each version of python, install the prerequisite
packages, and run your unit tests.



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [0.3.0] - 2018-10-21
### Fixed
- Fixed bug due to how help works in cmd2

### Changed
- cmd2-submenu now requires cmd2 version 0.9.6+


## [0.2.0] - 2018-10-15
### Initial release to PyPI
- Version 0.1.0 had been "released" on GitHub but we had forgotten to publish it to PyPI

tests/conftest.py
#
# coding=utf-8
def normalize(block):
    """ Normalize a block of text to perform comparison.

    Strip newlines from the very beginning and very end  Then split into separate lines and strip trailing whitespace
    from each line.
    """
    assert isinstance(block, str)
    block = block.strip('\n')
    return [line.rstrip() for line in block.splitlines()]


def run_cmd(app, cmd):
    """ Clear StdSim buffer, run the command, extract the buffer contents, """
    app.stdout.clear()
    app.onecmd_plus_hooks(cmd)
    out = app.stdout.getvalue()
    app.stdout.clear()
    return normalize(out)


tests/test_submenu.py
#
# coding=utf-8
"""
Cmd2 testing for argument parsing
"""
import readline

import pytest
from unittest import mock

from cmd2 import cmd2
import cmd2_submenu
from conftest import run_cmd, normalize
from cmd2.utils import StdSim

class SecondLevelB(cmd2.Cmd):
    """To be used as a second level command class. """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.prompt = '2ndLevel B '

    def do_get_top_level_attr(self, line):
        self.poutput(str(self.top_level_attr))

    def do_set_top_level_attr(self, line):
        self.top_level_attr = 987654321


class SecondLevel(cmd2.Cmd):
    """To be used as a second level command class. """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.prompt = '2ndLevel '
        self.top_level_attr = None

    def do_say(self, line):
        self.poutput("You called a command in SecondLevel with '%s'. " % line)

    def help_say(self):
        self.poutput("This is a second level menu. Options are qwe, asd, zxc")

    def complete_say(self, text, line, begidx, endidx):
        return [s for s in ['qwe', 'asd', 'zxc'] if s.startswith(text)]

    def do_get_top_level_attr(self, line):
        self.poutput(str(self.top_level_attr))

    def do_get_prompt(self, line):
        self.poutput(self.prompt)


second_level_cmd = SecondLevel()
second_level_b_cmd = SecondLevelB()


@cmd2_submenu.AddSubmenu(SecondLevelB(),
                 command='should_work_with_default_kwargs')
@cmd2_submenu.AddSubmenu(second_level_b_cmd,
                 command='secondb',
                 shared_attributes=dict(top_level_attr='top_level_attr'),
                 require_predefined_shares=False,
                 preserve_shares=True
                 )
@cmd2_submenu.AddSubmenu(second_level_cmd,
                 command='second',
                 aliases=('second_alias',),
                 shared_attributes=dict(top_level_attr='top_level_attr'))
class SubmenuApp(cmd2.Cmd):
    """To be used as the main / top level command class that will contain other submenus."""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.prompt = 'TopLevel '
        self.top_level_attr = 123456789

    def do_say(self, line):
        self.poutput("You called a command in TopLevel with '%s'. " % line)

    def help_say(self):
        self.poutput("This is a top level submenu. Options are qwe, asd, zxc")

    def complete_say(self, text, line, begidx, endidx):
        return [s for s in ['qwe', 'asd', 'zxc'] if s.startswith(text)]


@pytest.fixture
def submenu_app():
    app = SubmenuApp()
    app.stdout = StdSim(app.stdout)
    second_level_cmd.stdout = StdSim(app.stdout)
    second_level_b_cmd.stdout = StdSim(app.stdout)
    return app


@pytest.fixture
def secondlevel_app():
    app = SecondLevel()
    app.stdout = StdSim(app.stdout)
    return app


@pytest.fixture
def secondlevel_app_b():
    app = SecondLevelB()
    app.stdout = StdSim(app.stdout)
    return app

def run_submenu_cmd(app, second_level_app, cmd):
    """ Clear StdSim buffers, run the command, extract the buffer contents."""
    app.stdout.clear()
    second_level_app.stdout.clear()
    app.onecmd_plus_hooks(cmd)
    out1 = app.stdout.getvalue()
    out2 = second_level_app.stdout.getvalue()
    app.stdout.clear()
    second_level_app.stdout.clear()
    return normalize(out1), normalize(out2)

def complete_tester(text, line, begidx, endidx, app):
    """
    This is a convenience function to test cmd2.complete() since
    in a unit test environment there is no actual console readline
    is monitoring. Therefore we use mock to provide readline data
    to complete().

    :param text: str - the string prefix we are attempting to match
    :param line: str - the current input line with leading whitespace removed
    :param begidx: int - the beginning index of the prefix text
    :param endidx: int - the ending index of the prefix text
    :param app: the cmd2 app that will run completions
    :return: The first matched string or None if there are no matches
             Matches are stored in app.completion_matches
             These matches also have been sorted by complete()
    """
    def get_line():
        return line

    def get_begidx():
        return begidx

    def get_endidx():
        return endidx

    first_match = None
    with mock.patch.object(readline, 'get_line_buffer', get_line):
        with mock.patch.object(readline, 'get_begidx', get_begidx):
            with mock.patch.object(readline, 'get_endidx', get_endidx):
                # Run the readline tab-completion function with readline mocks in place
                first_match = app.complete(text, 0)

    return first_match

######
#
# test submenu functionality
#
######
def test_submenu_say_from_top_level(submenu_app):
    line = 'testing'
    out1, out2 = run_submenu_cmd(submenu_app, second_level_cmd, 'say ' + line)
    assert len(out1) == 1
    assert len(out2) == 0
    assert out1[0] == "You called a command in TopLevel with {!r}.".format(line)

def test_submenu_second_say_from_top_level(submenu_app):
    line = 'testing'
    out1, out2 = run_submenu_cmd(submenu_app, second_level_cmd, 'second say ' + line)

    # No output expected from the top level
    assert out1 == []

    # Output expected from the second level
    assert len(out2) == 1
    assert out2[0] == "You called a command in SecondLevel with {!r}.".format(line)

def test_submenu_say_from_second_level(secondlevel_app):
    line = 'testing'
    out = run_cmd(secondlevel_app, 'say ' + line)
    assert out == ["You called a command in SecondLevel with '%s'." % line]

def test_submenu_help_second_say_from_top_level(submenu_app):
    out1, out2 = run_submenu_cmd(submenu_app, second_level_cmd, 'help second say')
    # No output expected from the top level
    assert out1 == []

    # Output expected from the second level
    assert out2 == ["This is a second level menu. Options are qwe, asd, zxc"]

def test_submenu_help_say_from_second_level(secondlevel_app):
    out = run_cmd(secondlevel_app, 'help say')
    assert out == ["This is a second level menu. Options are qwe, asd, zxc"]

def test_submenu_help_second(submenu_app):
    out1, out2 = run_submenu_cmd(submenu_app, second_level_cmd, 'help second')
    out3 = run_cmd(second_level_cmd, 'help')
    assert out2 == out3

def test_submenu_from_top_help_second_say(submenu_app):
    out1, out2 = run_submenu_cmd(submenu_app, second_level_cmd, 'help second say')
    out3 = run_cmd(second_level_cmd, 'help say')
    assert out2 == out3

def test_submenu_shared_attribute(submenu_app):
    out1, out2 = run_submenu_cmd(submenu_app, second_level_cmd, 'second get_top_level_attr')
    assert out2 == [str(submenu_app.top_level_attr)]

def test_submenu_shared_attribute_preserve(submenu_app):
    out1, out2 = run_submenu_cmd(submenu_app, second_level_b_cmd, 'secondb get_top_level_attr')
    assert out2 == [str(submenu_app.top_level_attr)]
    out1, out2 = run_submenu_cmd(submenu_app, second_level_b_cmd, 'secondb set_top_level_attr')
    assert submenu_app.top_level_attr == 987654321
    out1, out2 = run_submenu_cmd(submenu_app, second_level_b_cmd, 'secondb get_top_level_attr')
    assert out2 == [str(987654321)]


######
#
# test completion in submenus
#
######

def test_cmd2_submenu_completion_single_end(submenu_app):
    text = 'sa'
    line = 'second {}'.format(text)
    endidx = len(line)
    begidx = endidx - len(text)

    first_match = complete_tester(text, line, begidx, endidx, submenu_app)

    # It is at end of line, so extra space is present
    assert first_match is not None and submenu_app.completion_matches == ['say ']

def test_cmd2_submenu_completion_multiple(submenu_app):
    text = 'e'
    line = 'second {}'.format(text)
    endidx = len(line)
    begidx = endidx - len(text)

    expected = ['edit', 'eof', 'eos']
    first_match = complete_tester(text, line, begidx, endidx, submenu_app)

    assert first_match is not None and submenu_app.completion_matches == expected

def test_cmd2_submenu_completion_nomatch(submenu_app):
    text = 'z'
    line = 'second {}'.format(text)
    endidx = len(line)
    begidx = endidx - len(text)

    first_match = complete_tester(text, line, begidx, endidx, submenu_app)
    assert first_match is None

def test_cmd2_submenu_completion_after_submenu_match(submenu_app):
    text = 'a'
    line = 'second say {}'.format(text)
    endidx = len(line)
    begidx = endidx - len(text)

    first_match = complete_tester(text, line, begidx, endidx, submenu_app)
    assert first_match is not None and submenu_app.completion_matches == ['asd ']

def test_cmd2_submenu_completion_after_submenu_nomatch(submenu_app):
    text = 'b'
    line = 'second say {}'.format(text)
    endidx = len(line)
    begidx = endidx - len(text)

    first_match = complete_tester(text, line, begidx, endidx, submenu_app)
    assert first_match is None

def test_cmd2_help_submenu_completion_multiple(submenu_app):
    text = 'p'
    line = 'help second {}'.format(text)
    endidx = len(line)
    begidx = endidx - len(text)

    matches = sorted(submenu_app.complete_help(text, line, begidx, endidx))
    assert matches == ['py', 'pyscript']

def test_cmd2_help_submenu_completion_nomatch(submenu_app):
    text = 'fake'
    line = 'help second {}'.format(text)
    endidx = len(line)
    begidx = endidx - len(text)
    assert submenu_app.complete_help(text, line, begidx, endidx) == []

def test_cmd2_help_submenu_completion_subcommands(submenu_app):
    text = 'p'
    line = 'help second {}'.format(text)
    endidx = len(line)
    begidx = endidx - len(text)

    matches = sorted(submenu_app.complete_help(text, line, begidx, endidx))
    assert matches == ['py', 'pyscript']


tasks.py
#
# -*- coding: utf-8 -*-
"""Development related tasks to be run with 'invoke'"""

import os
import re
import shutil
import sys

import invoke

# shared function
def rmrf(items, verbose=True):
    "Silently remove a list of directories or files"
    if isinstance(items, str):
        items = [items]

    for item in items:
        if verbose:
            print("Removing {}".format(item))
        shutil.rmtree(item, ignore_errors=True)
        # rmtree doesn't remove bare files
        try:
            os.remove(item)
        except FileNotFoundError:
            pass


# create namespaces
namespace = invoke.Collection()
namespace_clean = invoke.Collection('clean')
namespace.add_collection(namespace_clean, 'clean')

#####
#
# pytest, tox, pylint, and codecov
#
#####
@invoke.task
def pytest(context):
    "Run tests and code coverage using pytest"
    context.run("pytest --cov=cmd2_submenu")
namespace.add_task(pytest)

@invoke.task
def pytest_clean(context):
    "Remove pytest cache and code coverage files and directories"
    #pylint: disable=unused-argument
    dirs = ['.pytest_cache', '.cache', '.coverage']
    rmrf(dirs)
namespace_clean.add_task(pytest_clean, 'pytest')

@invoke.task
def tox(context):
    "Run unit and integration tests on multiple python versions using tox"
    context.run("tox")
namespace.add_task(tox)

@invoke.task
def tox_clean(context):
    "Remove tox virtualenvs and logs"
    #pylint: disable=unused-argument
    rmrf('.tox')
namespace_clean.add_task(tox_clean, 'tox')

@invoke.task
def pylint(context):
    "Check code quality using pylint"
    context.run('pylint --rcfile=cmd2_submenu/pylintrc cmd2_submenu')
namespace.add_task(pylint)

@invoke.task
def pylint_tests(context):
    "Check code quality of test suite using pylint"
    context.run('pylint --rcfile=tests/pylintrc tests')
namespace.add_task(pylint_tests)


#####
#
# build and distribute
#
#####
BUILDDIR = 'build'
DISTDIR = 'dist'

@invoke.task
def build_clean(context):
    "Remove the build directory"
    #pylint: disable=unused-argument
    rmrf(BUILDDIR)
namespace_clean.add_task(build_clean, 'build')

@invoke.task
def dist_clean(context):
    "Remove the dist directory"
    #pylint: disable=unused-argument
    rmrf(DISTDIR)
namespace_clean.add_task(dist_clean, 'dist')

@invoke.task
def eggs_clean(context):
    "Remove egg directories"
    #pylint: disable=unused-argument
    dirs = set()
    dirs.add('.eggs')
    for name in os.listdir(os.curdir):
        if name.endswith('.egg-info'):
            dirs.add(name)
        if name.endswith('.egg'):
            dirs.add(name)
    rmrf(dirs)
namespace_clean.add_task(eggs_clean, 'eggs')

@invoke.task
def bytecode_clean(context):
    "Remove __pycache__ directories and *.pyc files"
    #pylint: disable=unused-argument
    dirs = set()
    for root, dirnames, files in os.walk(os.curdir):
        if '__pycache__' in dirnames:
            dirs.add(os.path.join(root, '__pycache__'))
        for file in files:
            if file.endswith(".pyc"):
                dirs.add(os.path.join(root,file))
    print("Removing __pycache__ directories and .pyc files")
    rmrf(dirs, verbose=False)
namespace_clean.add_task(bytecode_clean, 'bytecode')

#
# make a dummy clean task which runs all the tasks in the clean namespace
clean_tasks = list(namespace_clean.tasks.values())
@invoke.task(pre=list(namespace_clean.tasks.values()), default=True)
def clean_all(context):
    "Run all clean tasks"
    #pylint: disable=unused-argument
    pass
namespace_clean.add_task(clean_all, 'all')


@invoke.task
def tag(context, name, message=''):
    "Add a Git tag and push it to origin"
    # If a tag was provided on the command-line, then add a Git tag and push it to origin
    if name:
        context.run('git tag -a {} -m {!r}'.format(name, message))
        context.run('git push origin {}'.format(name))
namespace.add_task(tag)

@invoke.task()
def validatetag(context):
    "Check to make sure that a tag exists for the current HEAD and it looks like a valid version number"
    # Validate that a Git tag exists for the current commit HEAD
    result = context.run("git describe --exact-match --tags $(git log -n1 --pretty='%h')")
    tag = result.stdout.rstrip()

    # Validate that the Git tag appears to be a valid version number
    ver_regex = re.compile('(\d+)\.(\d+)\.(\d+)')
    match = ver_regex.fullmatch(tag)
    if match is None:
        print('Tag {!r} does not appear to be a valid version number'.format(tag))
        sys.exit(-1)
    else:
        print('Tag {!r} appears to be a valid version number'.format(tag))
namespace.add_task(validatetag)

@invoke.task(pre=[clean_all])
def sdist(context):
    "Create a source distribution"
    context.run('python setup.py sdist')
namespace.add_task(sdist)

@invoke.task(pre=[clean_all])
def wheel(context):
    "Build a wheel distribution"
    context.run('python setup.py bdist_wheel')
namespace.add_task(wheel)

@invoke.task(pre=[validatetag, sdist, wheel])
def pypi(context):
    "Build and upload a distribution to pypi"
    context.run('twine upload dist/*')
namespace.add_task(pypi)

@invoke.task(pre=[validatetag, sdist, wheel])
def pypi_test(context):
    "Build and upload a distribution to https://test.pypi.org"
    context.run('twine upload --repository-url https://test.pypi.org/legacy/ dist/*')
namespace.add_task(pypi_test)


cmd2_submenu/submenu.py
#
# coding=utf-8
import copy
import readline
from typing import List

from cmd2.rl_utils import rl_type, RlType


class AddSubmenu(object):
    """Conveniently add a submenu (Cmd-like class) to a Cmd

    e.g. given "class SubMenu(Cmd): ..." then

    @AddSubmenu(SubMenu(), 'sub')
    class MyCmd(cmd.Cmd):
        ....

    will have the following effects:
    1. 'sub' will interactively enter the cmdloop of a SubMenu instance
    2. 'sub cmd args' will call do_cmd(args) in a SubMenu instance
    3. 'sub ... [TAB]' will have the same behavior as [TAB] in a SubMenu cmdloop
       i.e., autocompletion works the way you think it should
    4. 'help sub [cmd]' will print SubMenu's help (calls its do_help())
    """

    class _Nonexistent(object):
        """
        Used to mark missing attributes.
        Disable __dict__ creation since this class does nothing
        """
        __slots__ = ()  #

    def __init__(self,
                 submenu,
                 command,
                 aliases=(),
                 reformat_prompt="{super_prompt}>> {sub_prompt}",
                 shared_attributes=None,
                 require_predefined_shares=True,
                 create_subclass=False,
                 preserve_shares=False,
                 persistent_history_file=None
                 ):
        """Set up the class decorator

        submenu (Cmd):              Instance of something cmd.Cmd-like

        command (str):              The command the user types to access the SubMenu instance

        aliases (iterable):         More commands that will behave like "command"

        reformat_prompt (str):      Format str or None to disable
            if it's a string, it should contain one or more of:
                {super_prompt}:     The current cmd's prompt
                {command}:          The command in the current cmd with which it was called
                {sub_prompt}:       The subordinate cmd's original prompt
            the default is "{super_prompt}{command} {sub_prompt}"

        shared_attributes (dict):   dict of the form {'subordinate_attr': 'parent_attr'}
            the attributes are copied to the submenu at the last moment; the submenu's
            attributes are backed up before this and restored afterward

        require_predefined_shares: The shared attributes above must be independently
            defined in the subordinate Cmd (default: True)

        create_subclass: put the modifications in a subclass rather than modifying
            the existing class (default: False)
        """
        self.submenu = submenu
        self.command = command
        self.aliases = aliases
        if persistent_history_file:
            self.persistent_history_file = os.path.expanduser(persistent_history_file)
        else:
            self.persistent_history_file = None

        if reformat_prompt is not None and not isinstance(reformat_prompt, str):
            raise ValueError("reformat_prompt should be either a format string or None")
        self.reformat_prompt = reformat_prompt

        self.shared_attributes = {} if shared_attributes is None else shared_attributes
        if require_predefined_shares:
            for attr in self.shared_attributes.keys():
                if not hasattr(submenu, attr):
                    raise AttributeError("The shared attribute '{attr}' is not defined in {cmd}. Either define {attr} "
                                         "in {cmd} or set require_predefined_shares=False."
                                         .format(cmd=submenu.__class__.__name__, attr=attr))

        self.create_subclass = create_subclass
        self.preserve_shares = preserve_shares

    def _get_original_attributes(self):
        return {
            attr: getattr(self.submenu, attr, AddSubmenu._Nonexistent)
            for attr in self.shared_attributes.keys()
        }

    def _copy_in_shared_attrs(self, parent_cmd):
        for sub_attr, par_attr in self.shared_attributes.items():
            setattr(self.submenu, sub_attr, getattr(parent_cmd, par_attr))

    def _copy_out_shared_attrs(self, parent_cmd, original_attributes):
        if self.preserve_shares:
            for sub_attr, par_attr in self.shared_attributes.items():
                setattr(parent_cmd, par_attr, getattr(self.submenu, sub_attr))
        else:
            for attr, value in original_attributes.items():
                if attr is not AddSubmenu._Nonexistent:
                    setattr(self.submenu, attr, value)
                else:
                    delattr(self.submenu, attr)

    def __call__(self, cmd_obj):
        """Creates a subclass of Cmd wherein the given submenu can be accessed via the given command"""
        def enter_submenu(parent_cmd, statement):
            """
            This function will be bound to do_<submenu> and will change the scope of the CLI to that of the
            submenu.
            """
            submenu = self.submenu
            original_attributes = self._get_original_attributes()
            history = _pop_readline_history()

            if self.persistent_history_file and rl_type != RlType.NONE:
                try:
                    readline.read_history_file(self.persistent_history_file)
                except FileNotFoundError:
                    pass

            try:
                # copy over any shared attributes
                self._copy_in_shared_attrs(parent_cmd)

                if statement.args:
                    # Remove the menu argument and execute the command in the submenu
                    submenu.onecmd_plus_hooks(statement.args)
                else:
                    if self.reformat_prompt is not None:
                        prompt = submenu.prompt
                        submenu.prompt = self.reformat_prompt.format(
                            super_prompt=parent_cmd.prompt,
                            command=self.command,
                            sub_prompt=prompt,
                        )
                    submenu.cmdloop()
                    if self.reformat_prompt is not None:
                        # noinspection PyUnboundLocalVariable
                        self.submenu.prompt = prompt
            finally:
                # copy back original attributes
                self._copy_out_shared_attrs(parent_cmd, original_attributes)

                # write submenu history
                if self.persistent_history_file and rl_type != RlType.NONE:
                    readline.write_history_file(self.persistent_history_file)
                # reset main app history before exit
                _push_readline_history(history)

        def complete_submenu(_self, text, line, begidx, endidx):
            """
            This function will be bound to complete_<submenu> and will perform the complete commands of the submenu.
            """
            submenu = self.submenu
            original_attributes = self._get_original_attributes()
            try:
                # copy over any shared attributes
                self._copy_in_shared_attrs(_self)

                # Reset the submenu's tab completion parameters
                submenu.allow_appended_space = True
                submenu.allow_closing_quote = True
                submenu.completion_header = ''
                submenu.display_matches = []
                submenu.matches_delimited = False
                submenu.matches_sorted = False

                return _complete_from_cmd(submenu, text, line, begidx, endidx)
            finally:
                # copy back original attributes
                self._copy_out_shared_attrs(_self, original_attributes)

                # Pass the submenu's tab completion parameters back up to the menu that called complete()
                _self.allow_appended_space = submenu.allow_appended_space
                _self.allow_closing_quote = submenu.allow_closing_quote
                _self.completion_header = submenu.completion_header
                _self.display_matches = copy.copy(submenu.display_matches)
                _self.matches_delimited = submenu.matches_delimited
                _self.matches_sorted = submenu.matches_sorted

        original_do_help = cmd_obj.do_help
        original_complete_help_command = cmd_obj.complete_help_command

        def help_submenu(_self, line):
            """
            This function will be bound to help_<submenu> and will call the help commands of the submenu.
            """
            tokens = line.split(None, 1)
            if tokens and (tokens[0] == self.command or tokens[0] in self.aliases):
                self.submenu.do_help(tokens[1] if len(tokens) == 2 else '')
            else:
                original_do_help(_self, line)

        def _complete_submenu_help(_self, text, line, begidx, endidx):
            """autocomplete to match help_submenu()'s behavior"""
            tokens = line.split(None, 1)
            if len(tokens) == 2 and (
                    not (not tokens[1].startswith(self.command) and not any(
                        tokens[1].startswith(alias) for alias in self.aliases))
            ):
                return self.submenu.complete_help(
                    text,
                    tokens[1],
                    begidx - line.index(tokens[1]),
                    endidx - line.index(tokens[1]),
                )
            else:
                return original_complete_help_command(_self, text, line, begidx, endidx)

        if self.create_subclass:
            class _Cmd(cmd_obj):
                do_help = help_submenu
                complete_help = _complete_submenu_help
        else:
            _Cmd = cmd_obj
            _Cmd.do_help = help_submenu
            _Cmd.complete_help = _complete_submenu_help

        # Create bindings in the parent command to the submenus commands.
        setattr(_Cmd, 'do_' + self.command, enter_submenu)
        setattr(_Cmd, 'complete_' + self.command, complete_submenu)

        # Create additional bindings for aliases
        for _alias in self.aliases:
            setattr(_Cmd, 'do_' + _alias, enter_submenu)
            setattr(_Cmd, 'complete_' + _alias, complete_submenu)
        return _Cmd


def _pop_readline_history(clear_history: bool=True) -> List[str]:
    """Returns a copy of readline's history and optionally clears it (default)"""
    # noinspection PyArgumentList
    if rl_type == RlType.NONE:
        return []

    history = [
        readline.get_history_item(i)
        for i in range(1, 1 + readline.get_current_history_length())
    ]
    if clear_history:
        readline.clear_history()
    return history


def _push_readline_history(history, clear_history=True):
    """Restores readline's history and optionally clears it first (default)"""
    if rl_type != RlType.NONE:
        if clear_history:
            readline.clear_history()
        for line in history:
            readline.add_history(line)


def _complete_from_cmd(cmd_obj, text, line, begidx, endidx):
    """Complete as though the user was typing inside cmd's cmdloop()"""
    from itertools import takewhile
    command_subcommand_params = line.split(None, 3)

    if len(command_subcommand_params) < (3 if text else 2):
        n = len(command_subcommand_params[0])
        n += sum(1 for _ in takewhile(str.isspace, line[n:]))
        return cmd_obj.completenames(text, line[n:], begidx - n, endidx - n)

    command, subcommand = command_subcommand_params[:2]
    n = len(command) + sum(1 for _ in takewhile(str.isspace, line))
    cfun = getattr(cmd_obj, 'complete_' + subcommand, cmd_obj.complete)
    return cfun(text, line[n:], begidx - n, endidx - n)


examples/example.py
#!/usr/bin/env python
# coding=utf-8
"""
Create a CLI with a nested command structure as follows. The commands 'second' and 'third' navigate the CLI to the scope
of the submenu. Nesting of the submenus is done with the cmd2.AddSubmenu() decorator.

      (Top Level)----second----->(2nd Level)----third----->(3rd Level)
        |                            |                          |
        ---> say                     ---> say                   ---> say
"""
import sys

from cmd2 import cmd2
import cmd2_submenu
from IPython import embed


class ThirdLevel(cmd2.Cmd):
    """To be used as a third level command class. """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.prompt = '3rdLevel '
        self.top_level_attr = None
        self.second_level_attr = None

    def do_say(self, line):
        print("You called a command in ThirdLevel with '%s'. "
              "It has access to top_level_attr: %s "
              "and second_level_attr: %s" % (line, self.top_level_attr, self.second_level_attr))

    def help_say(self):
        print("This is a third level submenu (submenu_ab). Options are qwe, asd, zxc")

    def complete_say(self, text, line, begidx, endidx):
        return [s for s in ['qwe', 'asd', 'zxc'] if s.startswith(text)]


@cmd2_submenu.AddSubmenu(ThirdLevel(),
                 command='third',
                 aliases=('third_alias',),
                 shared_attributes=dict(second_level_attr='second_level_attr', top_level_attr='top_level_attr'))
class SecondLevel(cmd2.Cmd):
    """To be used as a second level command class. """
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.prompt = '2ndLevel '
        self.top_level_attr = None
        self.second_level_attr = 987654321

    def do_ipy(self, arg):
        """Enters an interactive IPython shell.

        Run python code from external files with ``run filename.py``
        End with ``Ctrl-D`` (Unix) / ``Ctrl-Z`` (Windows), ``quit()``, '`exit()``.
        """
        banner = 'Entering an embedded IPython shell type quit() or <Ctrl>-d to exit ...'
        exit_msg = 'Leaving IPython, back to {}'.format(sys.argv[0])
        embed(banner1=banner, exit_msg=exit_msg)

    def do_say(self, line):
        print("You called a command in SecondLevel with '%s'. "
              "It has access to top_level_attr: %s" % (line, self.top_level_attr))

    def help_say(self):
        print("This is a SecondLevel menu. Options are qwe, asd, zxc")

    def complete_say(self, text, line, begidx, endidx):
        return [s for s in ['qwe', 'asd', 'zxc'] if s.startswith(text)]


@cmd2_submenu.AddSubmenu(SecondLevel(),
                 command='second',
                 aliases=('second_alias',),
                 shared_attributes=dict(top_level_attr='top_level_attr'))
class TopLevel(cmd2.Cmd):
    """To be used as the main / top level command class that will contain other submenus."""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.prompt = 'TopLevel '
        self.top_level_attr = 123456789

    def do_ipy(self, arg):
        """Enters an interactive IPython shell.

        Run python code from external files with ``run filename.py``
        End with ``Ctrl-D`` (Unix) / ``Ctrl-Z`` (Windows), ``quit()``, '`exit()``.
        """
        banner = 'Entering an embedded IPython shell type quit() or <Ctrl>-d to exit ...'
        exit_msg = 'Leaving IPython, back to {}'.format(sys.argv[0])
        embed(banner1=banner, exit_msg=exit_msg)

    def do_say(self, line):
        print("You called a command in TopLevel with '%s'. "
              "TopLevel has attribute top_level_attr=%s" % (line, self.top_level_attr))

    def help_say(self):
        print("This is a top level submenu. Options are qwe, asd, zxc")

    def complete_say(self, text, line, begidx, endidx):
        return [s for s in ['qwe', 'asd', 'zxc'] if s.startswith(text)]


if __name__ == '__main__':

    root = TopLevel()
    root.cmdloop()

